### PR TITLE
Wait for the native thread to die in Thread.Join() on Windows

### PR DESCRIPTION
**Pull request by ntherning**
This is the behavior of .NET. After this patch the code on Mono for Windows will make sure the underlying native thread of a runtime created thread has died before `Thread.Join()` returns.

This PR should make some of the tests in `WaitHandleTest` less flaky on Windows.

### CODE REVIEW

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -1837,6 +1837,13 @@ ves_icall_System_Threading_Thread_Join_internal(MonoThread *this_obj, int ms)
 	mono_error_set_pending_exception (&error);
 
 	if(ret==MONO_THREAD_INFO_WAIT_RET_SUCCESS_0) {
+#ifdef HOST_WIN32
```
Comment (vargaz) [inline mono/metadata/threads.c]: Doesn't mono_join_uninterrupted () already do this ?

Comment (kumpera) [inline mono/metadata/threads.c]: Shouldn't this use the same mechanics as mono_threads_join_threads ?

**Comment by luhenry** [comment]
@niklas why is it flaky if we don't do that? Isn't it less flaky because some cases of races have just less chances of happening because it's a bit slower?

File: mono/metadata/threads.c (hunk 1 above)
Comment (ntherning) [inline mono/metadata/threads.c]: @vargaz IIUC `mono_join_uninterrupted` waits for an event to be signalled in the other thread (in `mono_threads_signal_thread_handle` which is called by `unregister_thread`). When the event is signalled the native thread has not yet been fully terminated.

@kumpera I guess you mean doing something similar to https://github.com/mono/mono/blob/master/mono/metadata/threads.c#L5048? Is that entire sequence starting with `MONO_ENTER_GC_SAFE` required?

**Comment by ntherning** [comment]
@luhenry The tests in `WaitHandleTest` which checks for abandoned mutexes on Windows. On Windows the mutex state won't change to abandoned until the native thread has died completely. But Our `Thread.Join()` doesn't guarantee this so there's a slight chance that `Join()` returns and the mutex state is checked before the other thread is completely dead. Those tests fail every now and then on Windows as you can see in Jenkins and locally. I cannot reproduce this failure on .NET so I assume that `Thread.Join()` in .NET doesn't return until the native thread of the other thread is actually terminated.

**Comment by ntherning** [comment]
build

**Comment by ntherning** [comment]
This appears to build fine and tests run as expected. Any objections to merging it? @kumpera You had a comment on the implementation?

**Comment by ntherning** [comment]
@kumpera @luhenry Please comment if there's anything I need to change in this PR. I'll merge this later this week if I hear no objections.

File: mono/metadata/threads.c (hunk 1 above)
Comment (kumpera) [inline mono/metadata/threads.c]: I mean using that very thing, but replacing pthread_join with WaitOne

**Comment by kumpera** [comment]
@ntherning I think we must first explore on whether extending the existing thread join code from unix to windows would not solve it.

We do it in this way to ensure proper lifetime of the underlying OS primitive and I'm not sure your approach allows for it.

**Comment by ntherning** [comment]
@kumpera One thing this PR addresses is the abandoned mutex issue which makes `WaitHandleTest.WaitAnyWithSecondMutexAbandoned` fail randomly. If we want to keep using real Windows mutexes I don't see how we can replicate the behavior of .NET without waiting for the underlying native thread to die in `Thread.Join()` like mono will do with this PR applied. To my knowledge there's no way to set a Windows mutex to abandoned without the owning native thread actually dying without releasing it.

The alternative to this PR would be to change Mutex on Windows to emulate mutexes using some other OS primitive and maintain the state ourselves, similarly to how it's done on unix. But that would break interoperability when managed code and native code are sharing mutexes. So not really an option.

**Comment by vargaz** [comment]
Wouldn't calling mono_native_thread_join () solve this ?

**Comment by ntherning** [comment]
`mono_native_thread_join()` does almost exactly what the PR does (calls `WaitForSingleObject()`). I can make that change.

I don't think this is what @kumpera had in mind though, but rather, if possible, align the thread joining on Windows with what we do on unix. As I tried to explain I don't think there's a way to do that while replicating what .NET does regarding mutexes.

**Comment by ntherning** [comment]
@kumpera I still need your comment on my previous [reply]( https://github.com/mono/mono/pull/5037#issuecomment-311580895).

**Comment by luhenry** [comment]
@ntherning if I understand correctly what @kumpera means is: please modify `mono_thread_join` / `mono_threads_join_threads` / `mono_threads_add_joinable_thread` to support win32, and use these functions, so we can use the same mechanism between Unix and Windows.

**Comment by ntherning** [comment]
But AFAICT `ves_icall_System_Threading_Thread_Join_internal()`, which this PR patches, doesn't use any of those functions you mention on any platform. This PR is about aligning the behavior of `Thread.Join()` on Mono on Windows with that of .NET, which will fix the problems we see in `MonoTests.System.Threading.WaitHandleTest.WaitAnyWithSecondMutexAbandoned`. I'm not trying to make other thread joins (internal to the runtime) behave like this on Windows. I don't think that is even desirable.

**Comment by kumpera** [comment]
Hi Niklas,

How about the following:

1) move threads.c:mono_join_uninterrupted to the w32 layer (or mono-thread depending on fit).
2) keep the current code as is for the unix version
3) On the windows version, do a WaitOne on the actual thread code.

File: mono/metadata/threads.c [hunk 2]
```diff
@@ -1837,6 +1844,13 @@ ves_icall_System_Threading_Thread_Join_internal(MonoThread *this_obj, int ms)
...
+		if (is_runtime_thread) {
+			// The thread was created by the runtime. Make sure the underlying
+			// native thread has terminated before we return.
+			WaitForSingleObjectEx (thread->native_handle, INFINITE, FALSE);
```
Comment (kumpera) [inline mono/metadata/threads.c]: This is wrong, the WaitOne should repect the timeout of the icall.

**Comment by vargaz** [comment]
How about we call mono_native_thread_join () which should do the same ?

**Comment by kumpera** [comment]
@vargaz AFAICT, we can't do that cuz pthread_join would fail on detached threads.
This is one of the reasons we have this convoluted thing in place.

OTOH, we could merge this as is and queue cleanup work to follow it. In such case, we'd only need to address the join timeout issue.

**Comment by vargaz** [comment]
We don't detach threads anymore, i.e. don't call pthread_detach ().

**Comment by kumpera** [comment]
Then I guess it's about how much @ntherning wants to go down the rabbit role.

**Comment by vargaz** [comment]
Superseded by #5454.

**Pull request closed without merging**